File: compiler/bench/knight.fs
\ expected: 336000
\ Knight's tour - count valid knight moves

8 constant SIZE
create board SIZE SIZE * allot

: b@ ( x y -- val ) SIZE * + board + c@ ;
: b! ( val x y -- ) SIZE * + board + c! ;

: in-bounds? ( x y -- flag )
  dup 0>= swap SIZE < and
  swap dup 0>= swap SIZE < and and ;

\ Knight move offsets
create dx 2 , 1 , -1 , -2 , -2 , -1 , 1 , 2 ,
create dy 1 , 2 , 2 , 1 , -1 , -2 , -2 , -1 ,

: count-valid-moves ( x y -- count )
  0
  8 0 do
    2 pick dx i cells + @ +
    2 pick dy i cells + @ +
    2dup in-bounds? if
      2dup b@ 0= if rot 1+ rot rot then
    then
    2drop
  loop
  nip nip ;

: init-board ( -- )
  SIZE SIZE * 0 do 0 board i + c! loop ;

: main
  0
  1000 0 do
    init-board
    SIZE 0 do SIZE 0 do
      j i count-valid-moves +
    loop loop
  loop
  . cr ;

File: compiler/bench/knight.c
// expected: 336000
// Knight's tour - count valid knight moves
#include <stdio.h>
#include <string.h>

#define SIZE 8

char board[SIZE * SIZE];

int b(int x, int y) { return board[y * SIZE + x]; }
void bs(int val, int x, int y) { board[y * SIZE + x] = val; }

int in_bounds(int x, int y) {
    return x >= 0 && x < SIZE && y >= 0 && y < SIZE;
}

int dx[] = {2, 1, -1, -2, -2, -1, 1, 2};
int dy[] = {1, 2, 2, 1, -1, -2, -2, -1};

int count_valid_moves(int x, int y) {
    int count = 0;
    for (int i = 0; i < 8; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];
        if (in_bounds(nx, ny) && !b(nx, ny)) {
            count++;
        }
    }
    return count;
}

void init_board(void) {
    memset(board, 0, SIZE * SIZE);
}

int main(void) {
    long sum = 0;
    for (int iter = 0; iter < 1000; iter++) {
        init_board();
        for (int y = 0; y < SIZE; y++) {
            for (int x = 0; x < SIZE; x++) {
                sum += count_valid_moves(x, y);
            }
        }
    }
    printf("%ld\n", sum);
    return 0;
}
